src/library/scala/collection/immutable/HashMap.scala

```diff
@@ -547,6 +675,103 @@ private final class BitmapIndexedMapNode[K, +V](
     this
   }
 
+  /** Variant of `removed` which will perform mutation on only the top-level node (`this`), rather than return a new
+    * node
+    *
+    * Should only be called on root nodes, because shift is assumed to be 0
+    *
+    * @param key the key to remove
+    * @param originalHash the original hash of `key`
+    * @param keyHash the improved hash of `key`
+    */
+  def removeWithShallowMutations[V1 >: V](key: K, originalHash: Int, keyHash: Int): this.type  = {
+    val mask = maskFrom(keyHash, 0)
+    val bitpos = bitposFrom(mask)
+
+    if ((dataMap & bitpos) != 0) {
+      val index = indexFrom(dataMap, mask, bitpos)
+      val key0 = this.getKey(index)
+
+      if (key0 == key) {
+        if (this.payloadArity == 2 && this.nodeMap == 0) {
+          val newDataMap = dataMap ^ bitpos
+          if (index == 0) {
+            val newContent = Array[Any](this.content(2), this.content(3))
+            val newOriginalHashes = Array(originalHashes(1))
+            val newCachedJavaKeySetHashCode = improve(getHash(1))
```

Comment:
Unwieldy names again

Reply:
Hmm I changed the other names but this one seems logically consistent with other var's. new-Content, new-OriginalHashes, new-CachedJavaKeySetHashCode. 

Maybe we should just change the name of the class's `var cachedJavaKeySetHashCode: Int` to `keySetHashCode` so that all these derived names aren't so long.

Reply:
Shortened